### PR TITLE
M2P-464 Refactor Magento EE GiftCard code (p2)

### PR DESCRIPTION
**Pull request by dumega**
# Description
Refactor Magento EE GiftCard code

Fixes: https://boltpay.atlassian.net/browse/M2P-464

#changelog Refactor Magento EE GiftCard code

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: ThirdPartyModules/Magento/GiftCardAccount.php [hunk 1]
```diff
@@ -196,4 +196,127 @@ function boltMagentoGiftInvalidateCart() {
...
+        $giftCard,
+        $quote
+    ) {
+        if ($result || !$giftCard instanceof \Magento\GiftCardAccount\Model\Giftcardaccount) {
+            return $result;
+        }
+        
+        try {
+            // on subsequest validation calls from Bolt checkout
+            // try removing the gift card before adding it
```
Comment (danac-gs) [inline ThirdPartyModules/Magento/GiftCardAccount.php:314]: This comment is redundant, please remove or change it.

File: ThirdPartyModules/Magento/GiftCardAccount.php [hunk 2]
```diff
@@ -196,4 +196,127 @@ function boltMagentoGiftInvalidateCart() {
...
+            $giftCard->removeFromCart(true, $quote);
+        } catch (\Exception $e) {
+            // gift card not added yet
+        } finally {
+            $giftCard->addToCart(true, $quote);
```
Comment (danac-gs) [inline ThirdPartyModules/Magento/GiftCardAccount.php:291]: Should we try / catch / return false here as well?

**Review by danac-gs — CHANGES_REQUESTED** [review]

File: ThirdPartyModules/Magento/GiftCardAccount.php (hunk 1 above)
Comment (dumega) [reply ThirdPartyModules/Magento/GiftCardAccount.php]: Removed

File: ThirdPartyModules/Magento/GiftCardAccount.php (hunk 2 above)
Comment (dumega) [reply ThirdPartyModules/Magento/GiftCardAccount.php:291]: @danac-gs In this PR, we just refactor the existing codes. and for your concern, I do notice it as well, and shall we move to this task https://boltpay.atlassian.net/browse/M2P-468 for discussion?

Comment (danac-gs) [reply ThirdPartyModules/Magento/GiftCardAccount.php:291]: I was just noticing that, comparing to the below `filterRemovingGiftCardCode` method, one of then can throw and another one can't, and they should probably act consistently. Not sure what the right approach is, but the change is small and could be implemented here if we agree. cc @vitaliyreznikov

**Review by danac-gs — APPROVED** [review]

**Review by vitaliyreznikov — APPROVED** [review]

**Pull request merged**